agent/test/unit/com/thoughtworks/go/agent/service/AutoregistrationPropertiesTest.java

```diff
@@ -0,0 +1,101 @@
+import java.util.Properties;
+
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertThat;
+
+public class AutoregistrationPropertiesTest {
+
+    @Rule
+    public TemporaryFolder folder = new TemporaryFolder();
+    private File configFile;
+
+    @Before
+    public void setUp() throws IOException {
+        configFile = folder.newFile();
+    }
+
+    @Test
+    public void shouldReturnAgentAutoRegisterPropertiesIfPresent() throws Exception {
+        Properties properties = new Properties();
+
+        properties.put(AutoregistrationProperties.AGENT_AUTO_REGISTER_KEY, "foo");
+        properties.put(AutoregistrationProperties.AGENT_AUTO_REGISTER_RESOURCES, "foo, zoo");
+        properties.put(AutoregistrationProperties.AGENT_AUTO_REGISTER_ENVIRONMENTS, "foo, bar");
+        properties.put(AutoregistrationProperties.AGENT_AUTO_REGISTER_HOSTNAME, "agent01.example.com");
+        properties.store(new FileOutputStream(configFile), "");
+
+        AutoregistrationProperties reader = new AutoregistrationProperties(configFile);
+        assertThat(reader.agentAutoRegisterKey(), is("foo"));
+        assertThat(reader.agentAutoRegisterResources(), is("foo, zoo"));
+        assertThat(reader.agentAutoRegisterEnvironments(), is("foo, bar"));
+        assertThat(reader.agentAutoRegisterHostname(), is("agent01.example.com"));
+    }
+
+    @Test
+    public void shouldReturnEmptyStringIfPropertiesNotPresent() {
+        AutoregistrationProperties reader = new AutoregistrationProperties(configFile);
+        assertThat(reader.agentAutoRegisterKey().isEmpty(), is(true));
+        assertThat(reader.agentAutoRegisterResources().isEmpty(), is(true));
+        assertThat(reader.agentAutoRegisterEnvironments().isEmpty(), is(true));
+        assertThat(reader.agentAutoRegisterHostname().isEmpty(), is(true));
+    }
+
+    @Test
+    public void shouldScrubTheRegistrationKey() throws Exception {
+        String originalContents = "" +
+                "#\n" +
+                "# file autogenerated by chef, any changes will be lost\n" +
+                "#\n" +
+                "# the registration key\n" +
+                "agent.auto.register.key = some secret key\n" +
+                "\n" +
+                "# the resources on this agent\n" +
+                "agent.auto.register.resources = some,resources\n" +
+                "\n";
+        FileUtils.write(configFile, originalContents);
+
+        AutoregistrationProperties properties = new AutoregistrationProperties(configFile);
+        properties.scrubRegistrationKey();
+
+        String newContents = "" +
+                "#\n" +
+                "# file autogenerated by chef, any changes will be lost\n" +
+                "#\n" +
+                "# the registration key\n" +
+                "# The autoregister key has been intentionally replaced by Go as a security measure.\n" +
+                "agent.auto.register.key = ***** Replaced by Go *****\n" +
```

Comment:
The call to `register` is guarded by [`registerIfNecessary`](https://github.com/gocd/gocd/blob/b0dbaf0acfc8d4411c21773e8da56b7aefe88fe5/agent/src/com/thoughtworks/go/agent/service/SslInfrastructureService.java#L89-99).

The current implementation assumes registered implies that the agent has a GUID and the keystore has the correct certs.

So to answer your question, there is no concept of "re-register" or "un-register" that an agent can possibly initiate.